README.md
## nonstop
Nonstop is a bootstrapper/service host that self-updates when new packages that match its configured parameters become available in the nonstop-index.

## Configuration
Nonstop can be used either as a library or as a command line interface.

As a command, you can configure it via a `bootstrap.json` file and environment variables.

> Note: environment variables take precendent over settings in the file.

### Bootstrap file
Defaults are shown here. The `index` property provides information necessary to contact the package index. The `package` property allows you to specify filtering information for the package. You shouldn't need to set the `architecture` or `platform` as those are detected for you.

```json
{
    "index": {
      "host": "localhost",
      "api": "/api",
      "frequency": 300000,
      "port": 4444,
      "ssl": false,
      "token": ""
    },
    "package": {
      "architecture": detected,
      "branch": "",
      "build": "",
      "owner": "",
      "platform": detected,
      "project": "",
      "releaseOnly": false,
      "version": "",
      "files": "./downloads"
    },
    service: {
      name: sysInfo.name,
      host: {
        ip: '',
        name: ''
      },
      port: {
        local: 9090,
        public: 9090
      },
      tolerance: 5000,
      failures: 1
    }
  }
```

### Environment Variables
| Group | Variable | Default |
|-------|-------------|---------|
| __Index__ | | |
| | INDEX_HOST | `"localhost"` |
| | INDEX_API | `"api"` |
| | INDEX_FREQUENCY | `5000` |
| | INDEX_PORT | `4444` |
| | INDEX_SSL | `false` |
| | INDEX_TOKEN | `""` |
| __Package__ | | |
| | PACKAGE_OWNER | `` |
| | PACKAGE_PROJECT | `` |
| | PACKAGE_BRANCH | `` |
| | PACKAGE_BUILD | `` |
| | PACKAGE_VERSION | `` |
| | PACKAGE__RELEASE_ONLY | `` |
| | PACKAGE_ARCHITECTURE | detected |
| | PACKAGE_PLATFORM | detected |
| | PACKAGE_FILES | `"./downloads"` |
| __service__ | | |
| | SERVICE_NAME | `"service name"` |
| | SERVICE_HOST_NAME | `"service name"` |
| | SERVICE_HOST_IP | `"unspecified` |
| | SERVICE_PORT_LOCAL | `9090` |
| | SERVICE_PORT_PUBLIC | `9090` |
| | SERVICE_FAILURES | `1` |
| | SERVICE_TOLERANCE | `5000` |

## Boot file - boot.yaml|boot.json
nonstop expects a boot file to be contained in any package it downloads which will provide the instructions for how it should start the packaged application. The files can be written in either JSON or YAML syntax.

The boot file consists of two sections: the service boot command and an optional pre-boot command set. The boot command simply tells nonstop how to start the packaged service while the optional pre-boot command set gets fed to [drudgeon](https://github.com/LeanKit-Labs/drudgeon). Both the boot command and pre-boot commands are expressed using `drudgeon`'s command syntax since it has a flexible means of supporting command and command set variation across platforms.

> Note: these examples are super arbitrary and should not be used to infer how you would actually create steps for an actual thing.

__JSON__
```javascript
{
  "boot": "node ./src/index.js",
  "preboot": {
    "one": {
      "win32": "gulp check-windows",
      "*": "gulp check"
    },
    "two": "node prep"
  }
}
```

__YAML__
```yaml
boot: "node ./src/index.js",
preboot:
  one:
    win32: "gulp check-windows"
    *: "gulp check"
  two: "node prep"
```




CHANGELOG.md
## 0.1.*

### 0.1.9
 * Bug fix - incorrect argument format passed to new drudgeon version
 * Switch specs to chai and add sinon-chai
 * Bug fix - remove hard coded path in spec expectation

### 0.1.8
Bug fix - preboot commands were using drudgeon's API incorrectly.

### 0.1.7
Correct edge cases where service stops checking for newer versions.

### 0.1.6
Fix defect preventing fresh installations from downloading a package.

### 0.1.5

 * Support new package format
 * Send notifications to registry on service stop/stop/fail events
 * Add ability to control failure tolerance via configuration and API
 * Bug fix - control commands should start/stop/reset hosted service
 * Bug fix - changes to package filtering should reset hosted service

### 0.1.4
Correct path to resources for global installation.

### 0.1.3
Introduction of HTTP API and auto-registration with index.

### 0.1.2
Bug fix - incorrect hash passed to index client library causing packages to filter incorrectly.

### 0.1.1

 * Simplify configuration approach
 * Update dependencies


src/config.js
var _ = require( "lodash" );
var path = require( "path" );
var sysInfo = require( "./sysInfo.js" )();
var filterFn = require( "./filter.js" );
var defaultDownloadPath = path.resolve( "./downloads" );
var defaultInstallPath = path.resolve( "./installs" );

function getDefaults() {
	return {
		index: {
			host: "localhost",
			api: "/api",
			frequency: 5000,
			port: 4444,
			ssl: false,
			token: "test"
		},
		package: {
			architecture: sysInfo.arch,
			branch: undefined,
			build: undefined,
			owner: undefined,
			platform: sysInfo.platform,
			project: undefined,
			releaseOnly: false,
			version: undefined,
			files: defaultDownloadPath,
			os: {}
		},
		service: {
			name: sysInfo.name,
			host: {
				name: sysInfo.name,
				ip: "unspecified"
			},
			port: {
				local: 9090,
				public: 9090
			},
			failures: 1,
			tolerance: 5000
		}
	};
}

function buildRootUrl( cfg ) {
	return [
		( cfg.ssl ? "https" : "http" ),
		"://",
		cfg.host,
		":",
		cfg.port,
		cfg.api
	].join( "" );
}

function buildServiceUrl( cfg ) {
	return [
		"http://",
		cfg.host,
		":",
		cfg.port.public,
		"api"
	].join( "" );
}

function buildDownloadRoot( cfg ) {
	return [
		( cfg.ssl ? "https" : "http" ),
		"://",
		cfg.host,
		":",
		cfg.port,
		cfg.packages
	].join( "" );
}

function getConfiguration( custom ) {
	var defaults = getDefaults();
	var merged = _.merge( defaults, custom );
	var cfg = require( "configya" )( {
		defaults: merged,
		file: "./bootstrap.json"
	} );

	cfg.package.osName = cfg.package.os.name || "any";
	cfg.package.osVersion = cfg.package.os.version || "any";
	return {
		filter: filterFn( cfg.package ),
		index: cfg.index,
		package: cfg.package,
		service: cfg.service,
		downloads: defaultDownloadPath,
		installs: defaultInstallPath,
		serviceRoot: buildServiceUrl( cfg.service ),
		apiRoot: buildRootUrl( cfg.index ),
		downloadRoot: buildDownloadRoot( cfg.index )
	};
}

module.exports = getConfiguration;
